packages/core/contracts/financial-templates/common/FundingRateApplier.sol

```diff
@@ -48,11 +50,6 @@ contract FundingRateApplier is Lockable {
     // If another 1% fee is charged, the multiplier should be 1.01^2 (1.0201).
     FixedPoint.Unsigned public cumulativeFundingRateMultiplier;
 
-    // FundingRateApplier and FeePayer are both at the same inheritance level within the PerpetualPositionManager. As a
```

Comment:
I'm not sure the change you made to `Tesatable` works like this. `FundingRateApplier` is now an instance of `Testable` but you never called the constructor with the timer address; in fact, you removed this timer address as a constructor param. As a result, I don't think `FundingRateApplier` is _testable_ with the same timer instance.

To validate this I pulled down your branch and ran tests on `PerpetualLiquidatable.js` and `PerpetualPositionManager.js`. All tests that modify and use the funding rate now fail with the funding rate not updating as expected (10 failing tests in total).

I am probably missing something here but it seems off to me. Also, CI agrees that these same tests failed.

If this implementation is correct and the timer works as expected works then it could be that the way the funding rate is pushed into the `FundingRateStore` in `PerpetualLiquidatable.js` and `PerpetualPositionManager.js` needs to be updated to accommodate centering around 0 as opposed to 1 how it is now. This is similar to the changes you made in the FundingRateApplier tests with `{ rawValue: toWei("1.0015") }, ->
      { rawValue: toWei("0.0015") }`

Reply:
Yeah, so I toyed around with it a little.

The idea is that you don't have to construct your parent classes _if you are declared abstract_ because abstract means you can't be constructed without something else deriving from you to fill in gaps in your impl. For constructors, you may rely on other classes in the inheritance tree to do the parent class construction, similar to how you might rely on another class to implement a method that you declared but didn't implement.

Does that make sense?

Like python, Solidity linearizes the inheritance tree under the hood to avoid duplicates or other diamond issues. Maybe this way of allowing abstract classes to not construct their base classes is something that's common in linearized inheritance languages...?

Reply:
ok cool! that's pretty awesome and something I didn't know about Solidity. The way you've got it set up now is _way_ cleaner than having a separate `testable` instance within the contracts :)

Reply:
@mrice32 so if the deriving contract is not `Testable` then the method calls would just revert?

Reply:
@nicholaspai I think the deriving contract is automatically `Testable` since it inherits a contract that is `Testable`. The weird part is that even the derived contract that isn't `Testable` can still call the `Testable` constructor in its constructor so its parent doesn't have to. This is what was done in `FundingRateApplierTest.sol`.